design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+### Transient queries -- Non-streaming
+
+The `Client` interface will also provide the following methods for receiving the results of a transient query (push or pull) in a single batch (non-streaming),
+once the query has completed:
+```
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql);
+
+  /**
+   * Execute a query (push or pull) and receive all result rows together, once the query has
+   * completed.
+   *
+   * @param sql statement of query to execute.
+   * @param properties query properties.
+   * @return query result.
+   */
+  CompletableFuture<List<Row>> executeQuery(String sql, Map<String, Object> properties);
+```
+
+For a query to "complete" could mean:
+* The query is a pull query
+* The query is a push query with a limit clause, and the limit has been reached
+* The query is a push query that has been terminated
+
+We may want to introduce a limit to the number of rows that may be returned from these `executeQuery()` methods,
+in order to decrease the likelihood of running out of memory.
+
+### Insert values
+
+A method to insert one row at a time:
+```
+  /**
+   * Insert a single row into the relevant stream/table.
+   *
+   * @param streamName name of stream/table.
+   * @param row the row to insert.
+   * @return a future that completes once the request has been processed.
+   */
+  CompletableFuture<Void> insertInto(String streamName, Map<String, Object> row);
+```
+
+A method to stream inserts (via the `/inserts-stream` endpoint):
+```
+  CompletableFuture<Publisher<InsertResponse>> streamInserts(String streamName, Publisher<List<Object>> insertsPublisher);
+}
+```
+where `InsertResponse` is as follows:
+```
+public interface InsertResponse {
+
+  /**
+   * Whether the row was successfully inserted or not.
+   */
+  boolean isSuccessful();
+
+  /**
+   * Unique sequence number for the row in the stream of inserts.
+   */
+  int getSequenceNum();
+
+  /**
+   * If unsuccessful, the error message.
+   */
+  String getErrorMessage();
+
+  /**
+   * If unsuccessful, the error code.
+   */
+  int getErrorCode();
+
+}
+```
+
+### DDL/DML operations
+
+#### `CREATE <STREAM/TABLE>`, `CREATE <STREAM/TABLE> ... AS SELECT`, `INSERT INTO`
+
+`Client` methods:
+```
+  /**
+   * Execute DDL statement: `CREATE STREAM`, `CREATE TABLE`
+   */
+  CompletableFuture<DdlResponse> executeDdl(String sql);
+
+  /**
+   * Execute DML statement: `CREATE STREAM ... AS SELECT`, `CREATE TABLE ... AS SELECT`, `INSERT INTO`
+   */
+  CompletableFuture<DmlResponse> executeDml(String sql);
+
+  /**
+   * Execute DML statement: `CREATE STREAM ... AS SELECT`, `CREATE TABLE ... AS SELECT`, `INSERT INTO`
+   */
+  CompletableFuture<DmlResponse> executeDml(String sql, Map<String, Object> properties);
```

Comment:
I wouldn't consider things like CTAS and CSAS as DML. DML in the relational DB world refers to operations which directly manipulate data, such as INSERT, DELETE and UPDATE. DDL refers to creating tables, indexes etc. I'd consider CTAS/CSAS as DDL.

Reply:
Agree with @purplefox on this. DML is generally understood to mean `INSERT`, `UPDATE`, `DELETE`, `TRUNCATE` etc. DDL is basically everything else (`CREATE`, `DROP`, etc.). I think we should keep with convention on this.